### PR TITLE
feat(tracks): Synchronize track operations.

### PR DESCRIPTION
**Pull request by hristoterezov**
Depends on https://github.com/jitsi/lib-jitsi-meet/pull/2163

### CODE REVIEW

**Review by saghul — COMMENTED** [review]
I took a look at it and left some comments / observations.

Generally speaking I am not comfortable with these changes.

First, I see no explanation as to WHAT this is solving. I do see, sort of, the HOW, making sure some operations run serialized, but it's very hard to tell what is being fixed in a 1000 line diff.

The current implementation requires that every new action dispatched which touches the track in a certain way uses the new stuff, but I am pretty sure new code is going to miss those, because the implementation is in the outermost layer, so the developer needs to be aware of this.

I think that design style is bug prone. WebRTC PeerConnection has an internal promise chain for example, but the user doesn't interact with it explicitly, operations are implicitly serialized.

Moreover, the fact mobile is not being considered here is very concerning IMHO. What reasons do we have to believe one platform needs it and the other doesn't? Even if it isn't necessary, what benefit do we get from having a different implementation on mobile? We try very hard to keep the gap as narrow as possible, as a general rule.

At a glance I think something like redux-saga might be a worthy candidate for consideration here.

Happy to discuss further, these are my first 2 cents!

**Comment by hristoterezov** [comment]
> I took a look at it and left some comments / observations.
> 
> Generally speaking I am not comfortable with these changes.
> 
> First, I see no explanation as to WHAT this is solving. I do see, sort of, the HOW, making sure some operations run serialized, but it's very hard to tell what is being fixed in a 1000 line diff.
> 

I'm sorry for the large PR. Unfortunately I don't see how to split it since I need to add almost all track operations in the promise chain. The problems that we solve here are the scenarios of the type - we do GUM and after it is resolve we do something with the new tracks ( for example the function for device handling) and in the same time some another operation that sneaks in between the promises changes the tracks( for example replaces tracks ).


> The current implementation requires that every new action dispatched which touches the track in a certain way uses the new stuff, but I am pretty sure new code is going to miss those, because the implementation is in the outermost layer, so the developer needs to be aware of this.
> 

Yes this is definitely disadvantage but the problems we are solving are exactly multi step operations that need to be atomic. 

> I think that design style is bug prone. WebRTC PeerConnection has an internal promise chain for example, but the user doesn't interact with it explicitly, operations are implicitly serialized.
> 

I'll be happy to hear more about this. Also I'll be happy to hear ideas for better design . 

> Moreover, the fact mobile is not being considered here is very concerning IMHO. What reasons do we have to believe one platform needs it and the other doesn't? Even if it isn't necessary, what benefit do we get from having a different implementation on mobile? We try very hard to keep the gap as narrow as possible, as a general rule.
> 

We discussed this with you a few weeks ago. As you mentioned the PR is quite big. Currently this PR resolves critical web issues that have been reported. We also aren't currently aware of the same issues beeing reproduced on mobile. For all those reasons this PR fixes the web ise case only. As we have discussed offline next steps will be to implement the track operations synchronization on mobile as well in a follow up PR. 

> At a glance I think something like redux-saga might be a worthy candidate for consideration here.
> 

I don't understand what you mean here. 


> Happy to discuss further, these are my first 2 cents!

Sure! Let's discuss offline since this feels like too long discussion.!

File: react/features/base/media/actions.ts [hunk 1]
```diff
@@ -53,10 +58,32 @@ export function setAudioAvailable(available: boolean) {
...
+    return (dispatch: IStore['dispatch'], getState: IStore['getState']) => {
+        const state = getState();
+        const participant = getLocalParticipant(state);
+
+        if (!muted && isForceMuted(participant, MEDIA_TYPE.AUDIO, state)) {
+            return Promise.resolve();
+        }
+
+        if (!muted
+            && isUserInteractionRequiredForUnmute(state)) {
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/base/media/actions.ts:70]: That seems correct!

I will open another PR that removes the function from everywhere. Is this OK?

File: react/features/base/media/actions.ts [hunk 2]
```diff
@@ -163,7 +205,7 @@ export function setVideoAvailable(available: boolean) {
  */
 export function setVideoMuted(
         muted: boolean,
-        mediaType: string = MEDIA_TYPE.VIDEO,
+        mediaType: MediaType = MEDIA_TYPE.VIDEO,
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/base/media/actions.ts:208]: This is correct!

I'd propose to remove this in a separate PR as well in order to not make this one even bigger. WDYT?

File: react/features/base/media/middleware.js [hunk 3]
```diff
@@ -234,9 +208,9 @@ function _setRoom({ dispatch, getState }, next, action) {
     // Unconditionally express the desires/expectations/intents of the app and
     // the user i.e. the state of base/media. Eventually, practice/reality i.e.
     // the state of base/tracks will or will not agree with the desires.
-    dispatch(setAudioMuted(audioMuted));
+    dispatch(executeTrackOperation(TrackOperationType.Audio, () => dispatch(setAudioMuted(audioMuted))));
     dispatch(setCameraFacingMode(CAMERA_FACING_MODE.USER));
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/base/media/middleware.js:212]: It seems to me it doesn't need to be in the queue, isn't it?

When I follow the code I don't see any promises been executed that could get mixed up with the other track operations.

Am I wrong? Maybe I'm missing something?

File: react/features/base/media/middleware.js [hunk 4]
```diff
@@ -296,21 +270,30 @@ function _setRoom({ dispatch, getState }, next, action) {
...
-    // fired before track gets to state.
-    if (track.muted !== muted) {
-        sendAnalytics(createSyncTrackStateEvent(mediaType, muted));
-        logger.log(`Sync ${mediaType} track muted state to ${muted ? 'muted' : 'unmuted'}`);
-
-        track.muted = muted;
-        setTrackMuted(track.jitsiTrack, muted, state);
-    }
+function _syncTrackMutedState({ dispatch, getState }, track) {
+    const mediaType = track.mediaType === MEDIA_TYPE.PRESENTER
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/base/media/middleware.js:274]: Wow. I think this is bad merge of conflicts.... Sorry! will fix it!

File: react/features/base/tracks/actions.any.ts [hunk 5]
```diff
@@ -817,37 +820,48 @@ export function updateLastTrackVideoMediaEvent(track: any, name: string): {
...
+            if (jitsiTrack && (
+                jitsiTrack.videoType !== 'desktop' || isAudioOnly || getMultipleVideoSendingSupportFeatureFlag(state))
+            ) {
+                return setTrackMuted(jitsiTrack, muted, state).catch(
+                    () => dispatch(trackMuteUnmuteFailed(localTrack, muted)));
+            }
+        } else if (!muted && ensureTrack && (typeof APP === 'undefined' || isPrejoinPageVisible(state))) {
+            // FIXME: This only runs on mobile now because web has its own way of
+            // creating local tracks. Adjust the check once they are unified.
+            return dispatch(createLocalTracksA({ devices: [ mediaType ] }));
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/base/tracks/actions.any.ts:862]: It actually does now!

But I've put the return of the promise in `createLocalTracksA` in the wrong place will have... Will fix this!

File: react/features/base/tracks/actions.web.ts [hunk 6]
```diff
@@ -216,7 +223,9 @@ async function _toggleScreenSharing(
             // Noise suppression doesn't work with desktop audio because we can't chain track effects yet, disable it
             // first. We need to to wait for the effect to clear first or it might interfere with the audio mixer.
             await dispatch(setNoiseSuppressionEnabled(false));
-            _maybeApplyAudioMixerEffect(desktopAudioTrack, state);
+
+            dispatch(executeTrackOperation(TrackOperationType.Audio,
+                () => _maybeApplyAudioMixerEffect(desktopAudioTrack, state)));
             dispatch(setScreenshareAudioTrack(desktopAudioTrack));
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/base/tracks/actions.web.ts:229]: Added!

File: react/features/base/tracks/actions.web.ts [hunk 7]
```diff
@@ -263,3 +281,102 @@ async function _toggleScreenSharing(
...
+/**
+ * Toggles the facingMode constraint on the video stream.
+ *
+ * @returns {Function}
+ */
+export function toggleCamera() {
+    return (dispatch: IStore['dispatch'], getState: IStore['getState']) =>
+        dispatch(executeTrackOperation(TrackOperationType.Video,
+
+            // eslint-disable-next-line valid-jsdoc
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/base/tracks/actions.web.ts:357]: removed!

**Comment by paweldomas** [comment]
I haven't read the entire PR and must be missing many thing, but I think the root cause of issues is that actually JitsiConference and other utils in lib-jitsi-meet are allowing for those problems to happen and the synchronization should happen there, not in jitsi-meet. This should cover the mobile app as well.

**Comment by paweldomas** [comment]
Specifically, JitsiConference  accepts calls to addTrack, replaceTrack etc. and operates on JingleSessionPC which has an async queue for those tasks, but there's lots of unprotected state on the JitsiConference level wrt to those calls. If those operations were queued correctly on the JitsiConference level, then jitsi-meet could rely on promises returned from those operations to execute things synchronously in jitsi-meet.

**Comment by hristoterezov** [comment]
@paweldomas maybe we have issues in LJM as well but here's what could happen in the client:
 - you unmute from a state where the track haven't been created at all.
 - you create the track
 - in mean time a device change is happening, which will basically try to set everything related to the tracks to point to the correct device
 - now the initial tracks from the unmutes are created and we go ahead and start using them with the wrong device.

The problem here mostly is that 1 user operation will include GUM, replaceTrack, muting tracks, etc. Most of those operations need correct and consistent state throughout the whole user operation. But because the separate LJM API calls are promises some other operation can sneak between the promises and change state which will be a big problem for the initial operation. Unless we change completely the LJM API to a higher level one where we have different methods for every user operation I don't think LJM can handle this internally because it just doesn't know what should be done. For example if we have 2 very fast user actions - first A and then B. And theoretically A includes creating a track and replacing the current one with the new one and B just calls track.mute() on the current track, we should do create new track, then replace the current track with the new track and then call track.mute. But if LJM receives them in the order create track, mute the current track and then replace the track, it would be impossible for LJM to know that the new track needs to be muted. 

Generally for many of those situations we can do workarounds like reading the state multiple times before each LJM operation or something else. But since we have so many of those cases it feels to me that we just need to consider them as atomic operation and queue them up. Otherwise every time you want to replace track, get the current track, etc you don't know if some previous operation is not waiting to be executed and replace your track with a wrong one.

Regarding the mobile app we can definitely implement it there the same way we do on web. Now the reason as I mentioned earlier to open the PR for web first is because we have reports for web-only about race conditions that are leaving the whole conference in a pretty bad state. AFAIK there aren't any such reports on mobile. Also this is a 1000 line PR with 30 files changed that I'm rebasing for 5th time. So it is better to merge this one first and then work on the mobile part. I expect that the mobile part will be easier because after this PR we have everything in place to start using it on mobile. And it is just a matter of analyzing the code and testing in order to find the critical places in the code for mobile where we need to use the synchronization mechanism.

**Comment by hristoterezov** [comment]
> > * you unmute from a state where the track haven't been created at all.
> > * you create the track
> > * in mean time a device change is happening, which will basically try to set everything related to the tracks to point to the correct device
> > * now the initial tracks from the unmutes are created and we go ahead and start using them with the wrong device.
> 
> Is this the only case we are fixing here? I don't think you can fix that really. A device change could still happen after gUM is called and before the promise is resolved, can't it?

This is not the only case. Other cases are: multiple initial unmute operations that are creating new track each, device change and quick unmute uses the wrong track and maybe some more. What become clear to me while working on the PR is that any 2 operations with tracks that are racing each other can lead to unpredictable results in the end.

On the second question: We are fixing the use case from my example with this PR. It will be fixed because the whole track creation and replace track operation will finish and then we will execute the whole device change handler and continue. They won't race each other because both operations are in the queue and resolve after the whole operation (unmute and device change handler) is completed.
 
> 
> > Generally for many of those situations we can do workarounds like reading the state multiple times before each LJM operation or something else. But since we have so many of those cases it feels to me that we just need to consider them as atomic operation and queue them up. Otherwise every time you want to replace track, get the current track, etc you don't know if some previous operation is not waiting to be executed and replace your track with a wrong one.
> 
> Yes, and this precisely why IMHO the current approach will fall short, because any operation need to be explicitly made to collaborate with the queue thing. So new code will likely miss things.

It is true that we need to be careful but I don't think this is a deal breaker. The other thing is that if we there is new operation that is not added to the queue, there is a big chance that we will notice some symptoms since all other operations are in the queue.

Also the other thing is that I don't have anything better mind in order to fix the problem.


> 
> > Regarding the mobile app we can definitely implement it there the same way we do on web. Now the reason as I mentioned earlier to open the PR for web first is because we have reports for web-only about race conditions that are leaving the whole conference in a pretty bad state. AFAIK there aren't any such reports on mobile.
> 
> I'd like to know more about the reports to see if I can correlate them with reports I've been hearing about.

Listed the reports a few lines above!

> 
> > Also this is a 1000 line PR with 30 files changed that I'm rebasing for 5th time. So it is better to merge this one first and then work on the mobile part.
> 
> How would you know you got the abstraction right if you ignored 50% of the supported platforms?

So I don't see what wouldn't fit into the abstraction since it accepts a function for the track operation. But anyway I think the more important part here is to decide if we scratch the PR, change it or we do something else.

> 
> > I expect that the mobile part will be easier because after this PR we have everything in place to start using it on mobile. And it is just a matter of analyzing the code and testing in order to find the critical places in the code for mobile where we need to use the synchronization mechanism.
> 
> You mean because all track management is in Redux?

This can be helpful indeed. The other thing is that we already have all the actions and logic for the queue implemented.

File: conference.js [hunk 8]
```diff
@@ -687,47 +668,55 @@ export default {
             });
     },
 
-    startConference(con, tracks) {
-        tracks.forEach(track => {
-            if ((track.isAudioTrack() && this.isLocalAudioMuted())
-                || (track.isVideoTrack() && this.isLocalVideoMuted())) {
-                const mediaType = track.getType();
+    async startConference(conP, tracksP, connectionAndTracksPromiseFunction) {
```
Comment (jallamsetty1) [inline conference.js:671]: Can you please add some JSDocs here ? Its hard to follow the logic and not very clear that this function either takes the connection and tracks objects or a promise that returns those two.

Comment (hristoterezov) [reply conference.js:684]: Done!

**Review by jallamsetty1 — COMMENTED** [review]
@hristoterezov I left few comments. Overall, there is still some inconsistency around when a track operation is pushed to the queue and when its not. Also, for some functions whole function body is executed on the track operation queue and in others multiple tasks are pushed on to the queue. What is the general principle that you are following and that others are supposed to follow ?

File: conference.js [hunk 9]
```diff
@@ -796,44 +785,55 @@ export default {
...
+                    let localTracks = handleStartAudioMuted(initialOptions, tracks);
 
-            // in case where gum is slow and resolves after the startAudio/VideoMuted coming from jicofo, we can be
-            // join unmuted even though jicofo had instruct us to mute, so let's respect that before passing the tracks
-            if (!browser.isWebKitBased()) {
-                if (room?.isStartAudioMuted()) {
-                    localTracks = localTracks.filter(track => track.getType() !== MEDIA_TYPE.AUDIO);
-                }
-            }
+                    // in case where gum is slow and resolves after the startAudio/VideoMuted coming from jicofo, we
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply conference.js:820]: So I don't think we should ignore this since if I understand correctly those tracks are going to be added after we have received the start muted from jicofo and therefore the newly added tracks won't be removed by LJM.

Otherwise I don't find moving the code for the mute in  `handleStartAudioMuted` related to the PR. But if you insist on it I'll move it. Do you want me to?

File: conference.js [hunk 10]
```diff
@@ -936,42 +936,45 @@ export default {
...
-            // This will only modify base/media.audio.muted which is then synced
-            // up with the track at the end of local tracks initialization.
-            muteLocalAudio(mute);
-            this.setAudioMuteStatus(mute);
+        APP.store.dispatch(executeTrackOperation(TrackOperationType.Audio, () => {
+            // Not ready to modify track's state yet
+            if (!this._localTracksInitialized) {
+                // This will only modify base/media.audio.muted which is then synced
+                // up with the track at the end of local tracks initialization.
+                const result = muteLocalAudio(mute);
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply conference.js:957]: definitely not intentional. Good catch! I probably lost that as well on one of the rebases.....

[the remark this answers is not in the document]
Comment (hristoterezov) [reply conference.js:957]: Will fix it!

File: conference.js [hunk 11]
```diff
@@ -936,42 +936,45 @@ export default {
...
+                        return null;
+                    })
+                    .then(async audioTrack => {
+                        await this._maybeApplyAudioMixerEffect(audioTrack);
+
+                        return this.useAudioStream(audioTrack);
+                    });
+            }
+
+            return muteLocalAudio(mute);
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply conference.js:989]: Will fix this!

**Comment by paweldomas** [comment]
I'm wondering if there are plans to go after synchronization problem at the lib-jitsi-meet in the future? Ultimately the problem is exactly on the `JitsiConference` level and no matter on how many other layers you'll fix it there will be always possible to trigger those bugs one way or the other. This fix on a fairly high layer compared to lib-jitsi-meet. So anything below this layer and JitsiConference the issue can still happen (adding new features etc).

**Comment by hristoterezov** [comment]
@paweldomas This totally makes sense! I'm actually not aware of any synchronization issues in LJM, could you be more specific? @jallamsetty1 do you know about any synchronization issues in LJM?

File: react/features/base/media/actions.ts [hunk 12]
```diff
@@ -53,10 +58,32 @@ export function setAudioAvailable(available: boolean) {
...
-        ensureTrack,
-        muted
+    return (dispatch: IStore['dispatch'], getState: IStore['getState']) => {
+        const state = getState();
+        const participant = getLocalParticipant(state);
+
+        if (!muted && isForceMuted(participant, MEDIA_TYPE.AUDIO, state)) {
+            return Promise.resolve();
+        }
+
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply react/features/base/media/actions.ts:68]: Hmmm... I moved this from one of the middlewares. Is this not valid anymore? 

To be honest I'm not sure what we actually do here. My goal was to keep the same logic and convert the simple action into an async one...

**Comment by hristoterezov** [comment]
> I left few comments. Overall, there is still some inconsistency around when a track operation is pushed to the queue and when its not. Also, for some functions whole function body is executed on the track operation queue and in others multiple tasks are pushed on to the queue. What is the general principle that you are following and that others are supposed to follow ?

It absolutely possible I have missed some places that need to be added in the queue because this thing is huuuugeeee :) Also tomorrow I plan to go over the PR one more time because obviously I have been victim on too many bad merges during rebase :) So if you are suspicious about anything please let me know! 

Otherwise I don't think we should introduce a general principle or rule. I think we should do that on case per case basis. When I was working on the PR I was trying to generally add to the queue small pieces that are as fast as possible and as simple as possible in order to not introduce delays and complexity. But also in the same in some cases this is not really possible or it is really complicated. For those cases I've included the whole function. Example: the device change handler.

IMHO the only harm of adding stuff to the queue would be if the queue gets slow. This would be my only worry there.

File: react/features/base/media/actions.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply react/features/base/media/actions.ts:68]: Well, I see that A/V moderation check is performed in base/media/actions for `setVideoMuted` and `setScreenshareMuted` but for audio it is performed in conference.js which is confusing.

[the remark this answers is not in the document]
Comment (jallamsetty1) [reply react/features/base/media/actions.ts:68]: Maybe we can fix this in a different PR.

File: react/features/base/tracks/actions.web.ts [hunk 13]
```diff
@@ -241,16 +256,25 @@ async function _toggleScreenSharing(
...
-            if (localAudio) {
-                localAudio.setEffect(undefined);
-            } else {
-                await conference.replaceTrack(desktopAudioTrack, null);
-            }
+            await dispatch(executeTrackOperation(TrackOperationType.Audio, async () => {
+                const localAudio = getLocalJitsiAudioTrack(state);
+
+                if (localAudio) {
+                    await localAudio.setEffect(undefined);
```
Comment (jallamsetty1) [inline react/features/base/tracks/actions.web.ts:273]: For posterity: as discussed, we need to check all other calls to track.setEffect() and add them to the track op queue as well since it involves replacing the track on the pc.

File: react/features/base/tracks/actions.any.ts [hunk 14]
```diff
@@ -216,7 +218,14 @@ export function createLocalTracksA(options: ITrackOptions = {}) {
                     mediaType: device
                 }
             });
+
+            promises.push(gumProcess.catch(() => {
+                // ignore the error in the result promises so that the Promise.all resolves after all promises are
```
Comment (saghul) [inline react/features/base/tracks/actions.any.ts:223]: Why not use Promise.allSettled then?
